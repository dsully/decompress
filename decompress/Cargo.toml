[package]
name = "decompress"
version = "0.6.0"
edition = "2021"

description = "Extracting archives made easy"
authors = ["Dotan Nahum <dotan@rng0.io>"]
documentation = "https://docs.rs/decompress/"
repository = "https://github.com/rusty-ferris-club/decompress"
keywords = ["compression", "archive", "unpack", "tar", "zip"]
license = "Apache-2.0"
readme = "../README.md"

[features]
default = ["all"]
all = [
  "tarball",
  "tarxz",
  "targz",
  "tarbz",
  "tarzst",
  "zip",
  "ar",
  "gz",
  "bz2",
  "xz",
  "zstd",
  "rar",
]

tarball = ["dep:tar"]
tarzst = ["tarball", "dep:zstd"]
targz = ["tarball", "dep:flate2"]
tarbz = ["tarball", "dep:bzip2"]
bz2 = ["dep:bzip2"]
tarxz = ["tarball", "dep:xz"]
zip = ["dep:zip"]
gz = ["dep:flate2"]
ar = ["dep:ar"]
xz = ["dep:xz"]
zstd = ["dep:zstd"]
rar = ["dep:unrar"]

[dependencies]

regex = "1.10.3"
lazy_static = "1.4.0"
thiserror = "1.0.57"
derive_builder = "0.20.0"
ar = { version = "0.9.0", optional = true }
tar = { version = "0.4.40", optional = true }
zip = { version = "0.6.6", optional = true }
bzip2 = { version = "0.4.4", optional = true }
flate2 = { version = "1.0.28", optional = true }
xz = { version = "0.1.0", optional = true }
zstd = { version = "0.13.0", optional = true }
unrar = { version = "0.5.3", optional = true }
infer = "0.15.0"

[dev-dependencies]
dircmp = "0.2.0"
rstest = "0.18.2"
clap = { version = "4.5.2", features = ["cargo"] }
insta = "1.36.1"

[[test]]
name = "archives_test"
required-features = ["all"]

[[example]]
name = "unpack"
required-features = ["default"]

[[example]]
name = "unzip"
required-features = ["zip"]
