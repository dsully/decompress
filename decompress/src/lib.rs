//! Decompress an archive, supporting multiple format and stripping path
//! prefixes
//!
//! ## Usage
//! You can use the default `decompress` stack:
//! ```ignore
#![doc = include_str!("../examples/unpack.rs")]
//!
//! ```
//! Or build your own stack:
//! ```ignore
#![doc = include_str!("../examples/unzip.rs")]
//!
//! ```
//! __NOTE__: to include or exclude [decompressors] types, use `--features` and disable default.
//! This in turn removes or includes the (costly) dependencies these features need.
//!
//!
//!
#![allow(clippy::missing_const_for_fn)]
#![allow(clippy::module_name_repetitions)]
pub mod decompressors;

use derive_builder::Builder;
use std::borrow::Cow;
use std::{convert::Infallible, io, path::Path};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum DecompressError {
    #[error("could not decompress: `{0}`")]
    IO(#[from] io::Error),

    #[error("could not decompress: `{0}`")]
    Error(String),

    #[error("could not decompress: `{0}`")]
    Infallible(#[from] Infallible),

    #[error("no compressor found")]
    MissingCompressor,
}

type FilterFn = dyn Fn(&Path) -> bool;
type MapFn = dyn Fn(&Path) -> Cow<'_, Path>;

#[derive(Builder)]
#[builder(pattern = "owned")]
pub struct ExtractOpts {
    #[builder(default)]
    pub strip: usize,

    #[builder(setter(custom), default = "Box::new(|_| true)")]
    pub filter: Box<FilterFn>,

    #[builder(setter(custom), default = "Box::new(|path| Cow::from(path))")]
    pub map: Box<MapFn>,
}

impl ExtractOptsBuilder {
    /// Given a predicate, filter a path in.
    #[must_use]
    pub fn filter(mut self, value: impl Fn(&Path) -> bool + 'static) -> Self {
        self.filter = Some(Box::new(value));
        self
    }
    /// Given a mapping function, transform a path into a different or similar path
    #[must_use]
    pub fn map(mut self, value: impl Fn(&Path) -> Cow<'_, Path> + 'static) -> Self {
        self.map = Some(Box::new(value));
        self
    }
}

#[derive(Debug)]
pub struct Decompression {
    pub id: &'static str,
}

///
/// `Decompressor` is a trait that you can implement to add your own decompressor type.
/// A `Decompressor` is inserted into a stack, where given a potential archive file,
/// many decompressors may attempt to test if they're capable of unpacking it.
/// The first `Decompressor` which will test true will be the one selected to unpack.
///
/// It is _recommended_ to let a user pick a regex for testing against a `Path`, although
/// there is no limit to what you can do, as long as a user can override the Decompressor
/// decision when building a custom stack.
///
pub trait Decompressor {
    ///
    /// Test if this `Decompressor` can unpack an archive, given a path.
    /// The convention is to use `Regex` internally to test a path, because this is
    /// a convenient way for end users to override the behavior.
    /// You may choose to implement a different, but configurable, testing strategy.
    fn test(&self, archive: &Path) -> bool;
    ///
    /// Decompress an archive
    ///
    /// # Errors
    ///
    /// This function will return an error if unpacking fails.
    fn decompress(
        &self,
        archive: &Path,
        to: &Path,
        opts: &ExtractOpts,
    ) -> Result<Decompression, DecompressError>;
}

///
/// Represent a stack of decompressors with a default stack preconfigured when calling `new`
///
pub struct Decompress {
    decompressors: Vec<Box<dyn Decompressor>>,
}

impl Default for Decompress {
    fn default() -> Self {
        Self {
            decompressors: vec![
                #[cfg(feature = "zip")]
                Box::<decompressors::zip::Zip>::default(),
                #[cfg(feature = "targz")]
                Box::<decompressors::targz::Targz>::default(),
                #[cfg(feature = "tarball")]
                Box::<decompressors::tarball::Tarball>::default(),
                #[cfg(feature = "tarxz")]
                Box::<decompressors::tarxz::Tarxz>::default(),
                #[cfg(feature = "tarbz")]
                Box::<decompressors::tarbz::Tarbz>::default(),
                #[cfg(feature = "tarzst")]
                Box::<decompressors::tarzst::Tarzst>::default(),
                // order is important, `gz` is placed only after the targz variant did not match
                // if it's placed above targz, it will unpack and leave a tar archive.
                #[cfg(feature = "gz")]
                Box::<decompressors::gz::Gz>::default(),
                #[cfg(feature = "ar")]
                Box::<decompressors::ar::Ar>::default(),
                #[cfg(feature = "bz2")]
                Box::<decompressors::bz2::Bz2>::default(),
                #[cfg(feature = "xz")]
                Box::<decompressors::xz::Xz>::default(),
                #[cfg(feature = "zstd")]
                Box::<decompressors::zstd::Zstd>::default(),
            ],
        }
    }
}

impl Decompress {
    /// Build given a custom stack of decompressors
    #[must_use]
    pub fn build(decompressors: Vec<Box<dyn Decompressor>>) -> Self {
        Self { decompressors }
    }

    /// Decompress
    ///
    /// # Errors
    ///
    /// This function will return an error if an IO or parsing error happened
    pub fn decompress<P: AsRef<Path>>(
        &self,
        archive: P,
        to: P,
        opts: &ExtractOpts,
    ) -> Result<Decompression, DecompressError> {
        if let Some(dec) = self
            .decompressors
            .iter()
            .find(|dec| dec.test(archive.as_ref()))
        {
            return dec.decompress(archive.as_ref(), to.as_ref(), opts);
        }
        Err(DecompressError::MissingCompressor)
    }

    /// Returns `true` if any of the decompressors in the stack can decompress this
    /// specific archive
    pub fn can_decompress<P: AsRef<Path>>(&self, archive: P) -> bool {
        self.decompressors
            .iter()
            .any(|dec| dec.test(archive.as_ref()))
    }
}

/// Decompress an archive with default decompressor set up
///
/// # Errors
///
/// This function will return an error if IO or parsing failed
pub fn decompress<P: AsRef<Path>>(
    archive: P,
    to: P,
    opts: &ExtractOpts,
) -> Result<Decompression, DecompressError> {
    Decompress::default().decompress(archive, to, opts)
}
