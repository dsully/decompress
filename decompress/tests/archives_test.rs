use std::{fs, path::Path};

use decompress::{decompressors, Decompress, DecompressError, Decompression, ExtractOptsBuilder};
use dircmp::Comparison;
use regex::Regex;
use rstest::rstest;

#[rstest]
#[case("inner.tar", "inner_0", 0, "tarball")]
#[case("bare.zip", "bare_zip_0", 0, "zip")]
#[case("bare.zip", "bare_zip_1", 1, "zip")]
#[case("bare.tar.gz", "bare_tgz_0", 0, "targz")]
#[case("bare.tar.gz", "bare_tgz_1", 1, "targz")]
#[case("bare.tar.xz", "bare_txz_0", 0, "tarxz")]
#[case("bare.tar.xz", "bare_txz_1", 1, "tarxz")]
#[case("folders.zip", "folders_zip_0", 0, "zip")]
#[case("folders.zip", "folders_zip_1", 1, "zip")]
#[case("folders.tar.gz", "folders_tgz_0", 0, "targz")]
#[case("folders.tar.gz", "folders_tgz_1", 1, "targz")]
#[case("folders.tar.xz", "folders_txz_0", 0, "tarxz")]
#[case("folders.tar.xz", "folders_txz_1", 1, "tarxz")]
#[case("inner.zip", "inner_zip_0", 0, "zip")]
#[case("inner.zip", "inner_zip_1", 1, "zip")]
#[case("inner.tar.gz", "inner_tgz_0", 0, "targz")]
#[case("inner.tar.gz", "inner_tgz_1", 1, "targz")]
#[case("inner.tar.xz", "inner_txz_0", 0, "tarxz")]
#[case("inner.tar.xz", "inner_txz_1", 1, "tarxz")]
#[case("inner.tar.zst", "inner_zst_1", 1, "tarzst")]
#[case("inner.tar.bz2", "inner_bz2_1", 1, "tarbz")]
#[case("bare.ar", "bare_ar", 0, "ar")]
#[case("sub.txt.gz", "gz_1", 0, "gz")]
#[case("sub.txt.bz2", "bz_2", 0, "bz2")]
#[case("sub.txt.xz", "xz_1", 0, "xz")]
#[case("sub.txt.zst", "zstd_1", 0, "zst")]
#[trace]
fn test_archives(
    #[case] archive: &str,
    #[case] outdir: &str,
    #[case] strip: usize,
    #[case] id: &str,
) {
    vec!["bare_zip_1", "bare_tgz_1", "bare_txz_1"]
        .iter()
        .map(|p| format!("tests/expected/{p}"))
        .for_each(|p| {
            if !Path::new(&p).exists() {
                let _res = fs::create_dir_all(&p);
            }
        });

    let extract_opts = ExtractOptsBuilder::default().strip(strip).build().unwrap();

    let res = assertion(archive, outdir, |from, to| {
        Decompress::default().decompress(from, to, &extract_opts)
    })
    .unwrap();

    assert_eq!(res.id, id);
}

#[test]
fn test_custom() {
    let extract_opts = ExtractOptsBuilder::default().build().unwrap();
    let dec = Decompress::build(vec![decompressors::targz::Targz::build(Some(
        Regex::new(r"(?i)\.tzz$").unwrap(),
    ))]);

    let res = assertion("tar-gz.tzz", "custom_tar_gz_tzz", |from, to| {
        dec.decompress(from, to, &extract_opts)
    })
    .unwrap();
    assert_eq!(res.id, "targz");

    // we swapped our decompressor stack, so now tar.gz should not work
    let res = assertion("bare.tar.gz", "bar_no_go", |from, to| {
        dec.decompress(from, to, &extract_opts)
    });

    match res {
        Err(DecompressError::MissingCompressor) => {}
        _ => panic!("should have not decompressed"),
    }
}

#[rstest]
#[case("bare.tar.gz", "bare_filter_tgz_0", "targz")]
#[case("bare.zip", "bare_filter_zip_0", "zip")]
#[trace]
fn test_filter(#[case] archive: &str, #[case] outdir: &str, #[case] id: &str) {
    let extract_opts = ExtractOptsBuilder::default()
        .strip(0)
        .filter(|path| {
            if let Some(path) = path.to_str() {
                return path.ends_with("ex.sh");
            }
            false
        })
        .build()
        .unwrap();

    let res = assertion(archive, outdir, |from, to| {
        Decompress::default().decompress(from, to, &extract_opts)
    })
    .unwrap();

    assert_eq!(res.id, id);
}

#[rstest]
#[case("bare.tar.gz", "bare_map_tgz_0", "targz")]
#[case("bare.zip", "bare_map_zip_0", "zip")]
#[trace]
fn test_map(#[case] archive: &str, #[case] outdir: &str, #[case] id: &str) {
    let extract_opts = ExtractOptsBuilder::default()
        .strip(0)
        .map(|path| {
            let mut path = path.to_path_buf();
            path.set_file_name(format!(
                "abc-{}",
                path.file_name().unwrap().to_str().unwrap()
            ));
            path.into()
        })
        .build()
        .unwrap();

    let res = assertion(archive, outdir, |from, to| {
        Decompress::default().decompress(from, to, &extract_opts)
    })
    .unwrap();

    assert_eq!(res.id, id);
}

#[test]
fn test_can_decompress() {
    assert!(Decompress::default().can_decompress("foo/bar/baz.tar.gz"));
    assert!(!Decompress::default().can_decompress("foo/bar/baz.tar.foo"));
}

fn assertion(
    from: &str,
    to: &str,
    extract: impl Fn(&str, &str) -> Result<Decompression, DecompressError>,
) -> Result<Decompression, DecompressError> {
    let from = format!("tests/fixtures/{from}");
    let out = format!("tests/out/{to}");

    if Path::new(&out).exists() {
        fs::remove_dir_all(&out).unwrap();
    }

    let extraction = extract(from.as_str(), out.as_str())?;

    let result = Comparison::default()
        .compare(Path::new(&out), Path::new(&format!("tests/expected/{to}")))
        .unwrap();

    assert!(result.is_empty());

    Ok(extraction)
}
