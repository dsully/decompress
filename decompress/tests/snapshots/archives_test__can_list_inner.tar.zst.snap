---
source: decompress/tests/archives_test.rs
expression: "(archive, Decompress::default().list(target))"
---
(
    "inner.tar.zst",
    Ok(
        Listing {
            id: "tarzst",
            entries: [
                "folder-1/",
                "folder-1/ex.sh",
                "folder-1/sub.txt",
            ],
        },
    ),
)
